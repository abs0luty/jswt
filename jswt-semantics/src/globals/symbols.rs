use crate::SemanticError;
use std::vec;

use jswt_ast::*;
use jswt_symbols::{ClassBinding, Field, FunctionBinding, SymbolTable};
use jswt_types::{PrimitiveType, Type};

#[derive(Debug)]
pub(crate) struct GlobalSymbolsResolver<'a> {
    pub symbols: &'a mut SymbolTable,
    pub stack: Vec<ClassBinding>,
    pub errors: Vec<SemanticError>,
}

impl<'a> GlobalSymbolsResolver<'a> {
    pub fn new(symbols: &'a mut SymbolTable) -> Self {
        Self {
            symbols,
            stack: vec![],
            errors: vec![],
        }
    }

    pub fn enter_program(&mut self, _: &Program) {
        self.symbols.push_global_scope();
    }

    pub fn exit_program(&mut self, _: &Program) {
        // we should only have the global scope
        debug_assert!(self.symbols.depth() == 1);
        self.symbols.pop_scope()
    }

    pub fn enter_function_declaration(&mut self, node: &FunctionDeclarationElement) {
        // add to global symbol table
        let ident = &node.ident;
        let name = &ident.value;

        let returns = node
            .returns
            .as_ref()
            .map(|it| it.ty.clone())
            .unwrap_or(Type::Void);

        let params = node
            .params
            .parameters
            .iter()
            .map(|param| param.type_annotation.ty.clone())
            .collect();

        if self.symbols.lookup_current(name).is_some() {
            let error = SemanticError::FunctionAlreadyDefined {
                name: name.clone(),
                span: ident.span.to_owned(),
            };
            self.errors.push(error);
        }

        self.symbols
            .define(name.clone(), FunctionBinding { params, returns });
    }

    pub(crate) fn exit_function_declaration(&mut self, _: &FunctionDeclarationElement) {
        // No-op
    }

    pub(crate) fn enter_constructor_declaration(&mut self, node: &ClassConstructorElement) {
        let binding = self.stack.last_mut().unwrap();
        for (i, param) in node.params.parameters.iter().enumerate() {
            binding.fields.push(Field {
                name: param.ident.value.clone(),
                index: i,
                size: 4,
            })
        }

        if self.symbols.lookup_current("constructor").is_some() {
            let error = SemanticError::FunctionAlreadyDefined {
                name: "constructor".into(),
                span: node.span.to_owned(),
            };
            self.errors.push(error);
        }

        // This should be the type binding but for now mark as a pointer
        let returns = Type::Primitive(PrimitiveType::I32);

        let params = node
            .params
            .parameters
            .iter()
            .map(|param| param.type_annotation.ty.clone())
            .collect();

        self.symbols
            .define("constructor", FunctionBinding { params, returns });
    }

    pub(crate) fn enter_class_declaration(&mut self, node: &ClassDeclarationElement) {
        let binding = ClassBinding {
            name: node.ident.value.clone(),
            fields: vec![],
        };
        self.stack.push(binding);
    }

    pub(crate) fn exit_class_declaration(&mut self, node: &ClassDeclarationElement) {
        let current = self.stack.pop().unwrap();
        let name = &node.ident.value;
        self.symbols.define(name.clone(), current);
    }

    pub(crate) fn enter_method_declaration(&self, node: &ClassMethodElement) {}
}
