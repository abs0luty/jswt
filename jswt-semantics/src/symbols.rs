use std::{borrow::Cow, collections::BTreeMap};

use jswt_common::Span;
use jswt_derive::FromEnumVariant;
use jswt_types::Type;

#[derive(Debug, Default)]
pub struct Scope {
    pub symbols: BTreeMap<Cow<'static, str>, Symbol>,
}

#[derive(Debug, PartialEq, FromEnumVariant)]
pub enum Symbol {
    Type(TypeBinding),
    Function(FunctionBinding),
    Class(ClassBinding),
}

impl Symbol {
    /// Returns `true` if the symbol is [`Function`].
    ///
    /// [`Function`]: Symbol::Function
    pub fn is_function(&self) -> bool {
        matches!(self, Self::Function(..))
    }
}

#[derive(Debug, PartialEq)]
pub struct TypeBinding {
    pub ty: Type,
}

#[derive(Debug, PartialEq)]
pub struct FunctionBinding {
    pub params: Vec<Type>,
    pub returns: Type,
}

#[derive(Debug, Default, PartialEq)]
pub struct ClassBinding {}

#[derive(Debug, Default)]
pub struct SymbolTable {
    scopes: Vec<Span>,
    table: BTreeMap<Span, Scope>,
}

impl SymbolTable {
    pub fn push_global_scope(&mut self) {
        // TODO - We want to have entities within the parser have
        // their own spans. We're inserting a synthetic one for now
        // just to make this work
        let global_span = Span::new("program".into(), "root".into(), 0, 0);
        self.push_scope(global_span);
    }

    // Pushing a scope adds it to the scope stack
    // and defines a new key in our global symbol map
    pub fn push_scope(&mut self, key: Span) {
        self.scopes.push(key.clone());
        self.table.entry(key).or_insert_with(Scope::default);
    }

    // Poping a scope removes it from the scope stack
    // but keeps the table reference around.
    // This allows future passes to find a symbol based on the span key
    // such that they don't have to do a full pass to find the symbol again
    pub fn pop_scope(&mut self) {
        self.scopes.pop();
    }

    // Returns the current size of the scopes stack
    // This is used to determine if we are in the global scope or not
    pub fn depth(&self) -> usize {
        self.scopes.len()
    }

    /// Define a symbol within the current active scope
    pub fn define<T: Into<Cow<'static, str>>, U: Into<Symbol>>(&mut self, name: T, symbol: U) {
        debug_assert!(!self.scopes.is_empty());
        let key = self.scopes.last().unwrap();
        let scope = self.table.get_mut(key).unwrap();
        scope.symbols.insert(name.into(), symbol.into());
    }

    /// Lookup a Symbol based on symbols in the current active scope
    /// as well as scopes lower in the stack
    pub fn lookup<T: Into<Cow<'static, str>>>(&mut self, name: T) -> Option<&Symbol> {
        // Apparently descending ranges aren't a thing.
        // as of 1.53, trying to use one doesn't generate a warning
        // https://github.com/rust-lang/rust/issues/70925
        let name = &name.into();
        for idx in (0..self.scopes.len()).rev() {
            let key = &self.scopes[idx];
            if let Some(scope) = self.table.get(key) {
                if let Some(sym) = scope.symbols.get(name) {
                    return Some(sym);
                }
            }
        }
        None
    }

    /// Look for the symbol in the local scope on
    /// the top of the stack
    pub fn lookup_current(&mut self, name: &str) -> Option<&Symbol> {
        debug_assert!(!self.scopes.is_empty());
        let key = &self.scopes.last().unwrap();
        self.table
            .get_mut(key)
            .and_then(|scope| scope.symbols.get(name))
    }

    /// Look for the symbol in the global scope on
    /// the top of the stack
    pub fn lookup_global(&mut self, name: &Cow<'static, str>) -> Option<&Symbol> {
        debug_assert!(!self.scopes.is_empty());
        let key = &self.scopes.first().unwrap();
        self.table
            .get_mut(key)
            .and_then(|scope| scope.symbols.get(name))
    }

    /// Returns all the symbols available in
    /// the current local scope
    pub fn get_scope(&mut self, key: Span) -> Option<&Scope> {
        self.table.get(&key)
    }
}
