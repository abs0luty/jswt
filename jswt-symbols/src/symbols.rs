use crate::{FunctionSignature, TypeSignature};
use jswt_common::{Span, Type};
use std::{borrow::Cow, collections::BTreeMap};

#[derive(Debug)]
pub enum Symbol {
    Type(TypeSignature),
    Function(FunctionSignature),
    Class,
    Unknown,
}

impl Symbol {
    pub fn ty(ty: Type) -> Self {
        Symbol::Type(TypeSignature { ty })
    }

    pub fn function(params: Vec<Type>, returns: Type) -> Self {
        Symbol::Function(FunctionSignature { params, returns })
    }
}

#[derive(Debug)]
pub struct Scope {
    pub symbols: BTreeMap<Cow<'static, str>, Symbol>,
    pub ret: Option<Type>,
}

impl Scope {
    pub fn new(ret: Option<Type>) -> Self {
        Self {
            symbols: BTreeMap::new(),
            ret,
        }
    }
}

#[derive(Debug, Default)]
pub struct ScopedSymbolTable {
    scopes: Vec<Span>,
    table: BTreeMap<Span, Scope>,
}

impl ScopedSymbolTable {
    // TODO - We want to have entities within the parser have
    // their own spans. We're inserting a synthetic one for now
    // just to make this work
    pub fn push_global_scope(&mut self) {
        self.push_scope(Span::new("program".into(), "root".into(), 0, 0), None);
    }

    // Pushing a scope adds it to the scope stack
    // and defines a new key in our global symbol map
    pub fn push_scope(&mut self, key: Span, returns: Option<Type>) {
        self.scopes.push(key.clone());
        self.table
            .entry(key.clone())
            .or_insert_with(|| Scope::new(returns));
    }

    // Poping a scope removes it from the scope stack
    // but keeps the table reference around.
    // This allows future passes to find a symbol based on the span key
    // such that they don't have to do a full pass to find the symbol again
    pub fn pop_scope(&mut self) {
        self.scopes.pop();
    }

    // Returns the current size of the scopes stack
    // This is used to determine if we are in the global scope or not
    pub fn depth(&self) -> usize {
        self.scopes.len()
    }

    /// Define a symbol within the current active scope
    pub fn define<T: Into<Cow<'static, str>>>(&mut self, name: T, symbol: Symbol) {
        debug_assert!(self.scopes.len() > 0);
        let key = self.scopes.last().unwrap();
        let scope = self.table.get_mut(key).unwrap();
        scope.symbols.insert(name.into(), symbol);
    }

    // pub fn update_type(&mut self, name: &Cow<'static, str>, ty: Type) {
    //     debug_assert!(self.scopes.len() > 0);
    //     let key = self.scopes.last().unwrap();
    //     let scope = self.table.get_mut(key).unwrap();
    //     if let Some(symbol) = scope.symbols.get_mut(name) {
    //         symbol.ty = ty;
    //     }
    // }

    /// Lookup a Symbol based on symbols in the current active scope
    /// as well as scopes lower in the stack
    pub fn lookup<T: Into<Cow<'static, str>>>(&mut self, name: T) -> Option<&Symbol> {
        // Apparently descending ranges aren't a thing.
        // as of 1.53, trying to use one doesn't generate a warning
        // https://github.com/rust-lang/rust/issues/70925
        let name = &name.into();
        for idx in (0..self.scopes.len()).rev() {
            let key = &self.scopes[idx];
            let scope = self.table.get(key).unwrap();
            if let Some(sym) = scope.symbols.get(name) {
                return Some(sym);
            }
        }
        None
    }

    /// Look for the symbol in the local scope on
    /// the top of the stack
    pub fn lookup_current(&mut self, name: &str) -> Option<&Symbol> {
        debug_assert!(self.scopes.len() > 0);
        let key = &self.scopes.last().unwrap();
        let scope = self.table.get_mut(key).unwrap();
        scope.symbols.get(name)
    }

    /// Look for the symbol in the global scope on
    /// the top of the stack
    pub fn lookup_global(&mut self, name: &Cow<'static, str>) -> Option<&Symbol> {
        debug_assert!(self.scopes.len() > 0);
        let key = &self.scopes.first().unwrap();
        let scope = self.table.get_mut(key).unwrap();
        scope.symbols.get(name)
    }

    /// Returns all the symbols available in
    /// the current local scope
    pub fn get_scope(&mut self, key: Span) -> Option<&Scope> {
        self.table.get(&key)
    }

    pub fn scope_return_type(&mut self) -> Option<&Type> {
        debug_assert!(self.scopes.len() > 0);
        let key = &self.scopes.last().unwrap();
        let scope = self.table.get_mut(key).unwrap();
        scope.ret.as_ref()
    }
}
