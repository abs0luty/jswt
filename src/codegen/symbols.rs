struct Scope {
    symbols: Vec<Symbol>,
}

struct Symbol {
    name: String,
}

struct SymbolTable {
    stack: Vec<Scope>,
}

impl SymbolTable {
    pub fn new() -> Self {
        Self { stack: vec![] }
    }

    pub fn push() {

    }
}
