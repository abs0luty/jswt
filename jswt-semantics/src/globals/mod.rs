mod symbols;

use jswt_ast::*;
use jswt_symbols::SymbolTable;

use self::symbols::*;
use crate::SemanticError;

#[derive(Debug)]
pub struct GlobalResolver<'a> {
    symbols: GlobalSymbolsResolver<'a>,
}

impl<'a> GlobalResolver<'a> {
    pub fn new(symbols: &'a mut SymbolTable) -> Self {
        Self {
            symbols: GlobalSymbolsResolver::new(symbols),
        }
    }

    pub fn resolve(&mut self, ast: &Ast) {
        self.visit_program(&ast.program);
    }

    pub fn errors(&mut self) -> Vec<SemanticError> {
        let mut errors = vec![];
        errors.append(&mut self.symbols.errors);
        errors
    }
}

impl<'a> ProgramVisitor<()> for GlobalResolver<'a> {
    fn visit_program(&mut self, node: &Program) {
        self.symbols.enter_program(node);
        for file in &node.files {
            self.visit_file(file)
        }
        self.symbols.exit_program(node);
    }

    fn visit_file(&mut self, node: &File) {
        self.visit_source_elements(&node.source_elements);
    }

    fn visit_source_elements(&mut self, node: &SourceElements) {
        for element in &node.source_elements {
            self.visit_source_element(element);
        }
    }

    fn visit_source_element(&mut self, node: &SourceElement) {
        match node {
            SourceElement::FunctionDeclaration(elem) => self.visit_function_declaration(elem),
            SourceElement::ClassDeclaration(elem) => self.visit_class_declaration(elem),
            SourceElement::Statement(elem) => self.visit_statement_element(elem),
        }
    }
}

impl<'a> StatementVisitor<()> for GlobalResolver<'a> {
    fn visit_statement_element(&mut self, node: &StatementElement) {
        match node {
            StatementElement::Block(stmt) => self.visit_block_statement(stmt),
            StatementElement::Empty(stmt) => self.visit_empty_statement(stmt),
            StatementElement::Return(stmt) => self.visit_return_statement(stmt),
            StatementElement::Variable(stmt) => self.visit_variable_statement(stmt),
            StatementElement::Expression(stmt) => self.visit_expression_statement(stmt),
            StatementElement::If(stmt) => self.visit_if_statement(stmt),
            StatementElement::Iteration(_) => {}
        };
    }

    fn visit_block_statement(&mut self, _: &BlockStatement) {
        // No-op
    }

    fn visit_empty_statement(&mut self, _: &EmptyStatement) {
        // No-op
    }

    fn visit_if_statement(&mut self, _: &IfStatement) {
        // No-op
    }

    fn visit_iteration_statement(&mut self, _: &IterationStatement) {
        unreachable!()
    }

    fn visit_while_iteration_element(&mut self, _: &WhileIterationElement) {
        unreachable!()
    }

    fn visit_return_statement(&mut self, _: &ReturnStatement) {
        // No-op
    }

    fn visit_variable_statement(&mut self, _: &VariableStatement) {
        // No-op
    }

    fn visit_expression_statement(&mut self, _: &ExpressionStatement) {
        // No-op
    }

    fn visit_statement_list(&mut self, node: &StatementList) {
        for statement in &node.statements {
            self.visit_statement_element(statement);
        }
    }

    fn visit_function_declaration(&mut self, node: &FunctionDeclarationElement) {
        self.symbols.enter_function_declaration(node);
        self.visit_block_statement(&node.body);
        self.symbols.exit_function_declaration(node);
    }

    fn visit_class_declaration(&mut self, node: &ClassDeclarationElement) {
        self.symbols.enter_class_declaration(node);
        self.visit_class_body(&node.body);
        self.symbols.exit_class_declaration(node);
    }

    fn visit_class_constructor_declaration(&mut self, node: &ClassConstructorElement) {
        self.symbols.enter_constructor_declaration(node);
    }

    fn visit_class_method_declaration(&mut self, node: &ClassMethodElement) {
        self.symbols.enter_method_declaration(node);
    }

    fn visit_class_body(&mut self, node: &ClassBody) {
        for element in &node.class_elements {
            match element {
                ClassElement::Constructor(elem) => self.visit_class_constructor_declaration(elem),
                ClassElement::Method(elem) => self.visit_class_method_declaration(elem),
            }
        }
    }
}

#[cfg(test)]
mod test {

    use super::*;
    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_tokenizer::Tokenizer;

    #[test]
    fn test_global_resolver_resolves_class_binding() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test.1",
            r"
        let global = 55;

        class Array {
            constructor(a: i32, b: f32) {}
            method() {}
        }

        function test() {
            let x = 99;
        }
 
        function doSomething(): Array { }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = SymbolTable::default();
        let mut resolver = GlobalResolver::new(&mut symbols);
        resolver.resolve(&ast);
        assert_debug_snapshot!(resolver);
    }

    #[test]
    fn test_global_resolver_resolves_function_bindings() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test.1",
            r"
        class Array {

        }

        function test(a: i32, b: i32) {
            let x = 99;
        }
 
        function test2(): Array {

        }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = SymbolTable::default();
        let mut resolver = GlobalResolver::new(&mut symbols);
        resolver.resolve(&ast);
        assert_debug_snapshot!(resolver);
    }
}
