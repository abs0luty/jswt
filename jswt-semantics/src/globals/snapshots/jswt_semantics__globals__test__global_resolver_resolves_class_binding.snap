---
source: jswt-semantics/src/globals/mod.rs
expression: resolver

---
GlobalResolver {
    symbols: GlobalSymbolsResolver {
        symbols: SymbolTable {
            scopes: [],
            table: {
                Span {
                    file: "program",
                    module: "root",
                    start: 0,
                    end: 0,
                }: Scope {
                    symbols: {
                        "Array": Class(
                            ClassBinding {
                                name: "Array",
                                fields: [
                                    Field {
                                        name: "a",
                                        index: 0,
                                        size: 4,
                                    },
                                    Field {
                                        name: "b",
                                        index: 1,
                                        size: 4,
                                    },
                                ],
                            },
                        ),
                        "constructor": Function(
                            FunctionBinding {
                                params: [
                                    Primitive(
                                        I32,
                                    ),
                                    Primitive(
                                        F32,
                                    ),
                                ],
                                returns: Primitive(
                                    I32,
                                ),
                            },
                        ),
                        "doSomething": Function(
                            FunctionBinding {
                                params: [],
                                returns: Object(
                                    Reference(
                                        "Array",
                                    ),
                                ),
                            },
                        ),
                        "test": Function(
                            FunctionBinding {
                                params: [],
                                returns: Void,
                            },
                        ),
                    },
                },
            },
        },
        stack: [],
        errors: [],
    },
}
